src/agent.py

```diff
@@ -68,31 +84,69 @@ def _on_agent_relation_joined(self, event: ops.RelationJoinedEvent) -> None:
             return
 
         logger.info("%s relation joined.", event.relation.name)
-        self.charm.unit.status = ops.MaintenanceStatus("Setting up relation.")
+        self.charm.unit.status = ops.MaintenanceStatus(
+            f"Setting up '{event.relation.name}' relation."
+        )
 
-        # handle slave relation for backwards compatibility.
-        if event.relation.name == SLAVE_RELATION:
-            logger.info(
-                "Slave relation data set: %s",
-                self.state.agent_meta.get_jenkins_slave_interface_dict(),
-            )
-            event.relation.data[self.charm.unit].update(
-                self.state.agent_meta.get_jenkins_slave_interface_dict()
+        relation_data = self.state.agent_meta.get_jenkins_agent_v0_interface_dict()
+        logger.info("Agent relation data set: %s", relation_data)
+        event.relation.data[self.charm.unit].update(relation_data)
+
+    def _on_slave_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        """Handle slave relation changed event.
+
+        Args:
+            event: The event fired when slave relation data has changed.
+        """
+        logger.info("%s relation changed.", event.relation.name)
+
+        if self.state.jenkins_config:
+            logger.warning(
+                "Jenkins configuration already exists. Ignoring %s relation.", event.relation.name
             )
             return
-        logger.info(
-            "Agent relation data set: %s",
-            self.state.agent_meta.get_jenkins_slave_interface_dict(),
+
+        container = self.charm.unit.get_container(self.state.jenkins_agent_service_name)
+        if not container.can_connect():
+            logger.warning("Jenkins agent container not yet ready. Deferring.")
+            event.defer()
+            return
+
+        if server.is_registered(container):
+            logger.warning("Given agent already registered. Skipping.")
+            return
+
+        credentials: typing.Optional[server.Credentials] = server.get_credentials(
+            event.relation.name,
+            unit_name=self.state.agent_meta.name,
+            databag=event.relation.data.get(typing.cast(ops.Unit, event.unit)),
         )
-        event.relation.data[self.charm.unit].update(
-            self.state.agent_meta.get_jenkins_agent_v0_interface_dict()
+        if not credentials:
+            self.charm.unit.status = ops.WaitingStatus("Waiting for complete relation data.")
+            logger.info("Waiting for complete relation data.")
+            return
+
+        self.charm.unit.status = ops.MaintenanceStatus("Validating credentials.")
+        if not server.validate_credentials(
+            agent_name=self.state.agent_meta.name,
+            credentials=credentials,
+            container=container,
+            add_random_delay=True,
```

Comment:
The default value is `False`, it's added for the special case for `slave_relation_changed` since this is the relation where all agents will try to register at once. The default `false` value is used for `agent_relation_changed` :D